### PR TITLE
Remove ambiguous fields from AuditEntry

### LINKED ISSUE #3016: Failed to unmarshal audit log entry
Using v57 of the package, we have a problem reading the audit log using `Organizations.GetAuditLog`. 

The `AuditEntry` struct has a field `Permission *string`, but we have found a scenario in which the API returns a bool in the permission field (see example below). This naturally leads to the error `cannot unmarshal bool into Go struct field AuditEntry.permission of type string`, which basically blocks us from reading the audit log.

```json
{
  "@timestamp": 1701766234196,
  "_document_id": "abc",
  "action": "org.update_member_repository_creation_permission",
  "actor": "ole-olsen",
  "actor_id": 123456789,
  "actor_location": {
    "country_code": "NO"
  },
  "business": "business",
  "business_id": 123456,
  "created_at": 1701766234196,
  "external_identity_nameid": "ole.olsen@a.com",
  "external_identity_username": "ole.olsen@a.com",
  "operation_type": "modify",
  "org": "myorg",
  "org_id": 123456789,
  "permission": true,
  "user_agent": "Mozilla/5.0 ...",
  "visibility": "public_internal"
}

### PR DESCRIPTION
**Pull request by WillAbides**
closes #3016

In https://github.com/google/go-github/issues/3016#issuecomment-1841102405 I suggested that AuditEntry fields should be limited to what is in the openapi description. I have since spoken with some people at GitHub and learned that the openapi description is outdated and there are discrepancies between the description and the json being returned by the api.

Instead of limiting AuditEntry to what is in the openapi description, this PR reduces AuditEntry to fields that are known to be the same type across all events regardless of the openapi description. The rest of the fields are now available in `AuditEntry.AdditionalFields` as a `map[string]any`

This is a breaking change.

### CODE REVIEW

File: github/orgs_audit_log.go [hunk 1]
```diff
@@ -34,103 +35,100 @@ type ActorLocation struct {
...
+
+	// All fields that are not explicitly defined in the struct are captured here.
+	AdditionalFields map[string]interface{} `json:"-"`
+}
+
+func (a *AuditEntry) UnmarshalJSON(data []byte) error {
+	type entryAlias AuditEntry
+	var v entryAlias
+	err := json.Unmarshal(data, &v)
+	if err != nil {
```
Comment (gmlewis) [inline github/orgs_audit_log.go:73]: nit

```suggestion
	if err := json.Unmarshal(data, &v); err != nil {
```

File: github/orgs_audit_log.go [hunk 2]
```diff
@@ -34,103 +35,100 @@ type ActorLocation struct {
...
+		return err
+	}
+
+	rawDefinedFields, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	definedFields := map[string]interface{}{}
+	err = json.Unmarshal(rawDefinedFields, &definedFields)
+	if err != nil {
```
Comment (gmlewis) [inline github/orgs_audit_log.go:83]:
```suggestion
	if err := json.Unmarshal(rawDefinedFields, &definedFields); err != nil {
```

File: github/orgs_audit_log.go [hunk 3]
```diff
@@ -34,103 +35,100 @@ type ActorLocation struct {
...
+		return err
+	}
+
+	err = json.Unmarshal(data, &v.AdditionalFields)
+	if err != nil {
```
Comment (gmlewis) [inline github/orgs_audit_log.go:88]:
```suggestion
	if err := json.Unmarshal(data, &v.AdditionalFields); err != nil {
```

File: github/orgs_audit_log.go [hunk 4]
```diff
@@ -34,103 +35,100 @@ type ActorLocation struct {
...
+		return err
+	}
+
+	for key, val := range v.AdditionalFields {
+		_, defined := definedFields[key]
+		if defined || val == nil {
```
Comment (gmlewis) [inline github/orgs_audit_log.go:94]:
```suggestion
		if _, ok := definedFields[key]; ok || val == nil {
```

File: github/orgs_audit_log.go [hunk 5]
```diff
@@ -34,103 +35,100 @@ type ActorLocation struct {
...
+	defBytes, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+	if len(a.AdditionalFields) == 0 {
+		return defBytes, err
+	}
+	resMap := map[string]interface{}{}
+	err = json.Unmarshal(defBytes, &resMap)
+	if err != nil {
```
Comment (gmlewis) [inline github/orgs_audit_log.go:118]:
```suggestion
	if err := json.Unmarshal(defBytes, &resMap); err != nil {
```

File: github/orgs_audit_log.go [hunk 6]
```diff
@@ -34,103 +35,100 @@ type ActorLocation struct {
...
+		return nil, err
+	}
+	for key, val := range a.AdditionalFields {
+		if val == nil {
+			continue
+		}
+		_, defined := resMap[key]
+		if defined {
```
Comment (gmlewis) [inline github/orgs_audit_log.go:126]:
```suggestion
		if _, ok := resMap[key]; ok {
```

File: github/orgs_audit_log.go [hunk 7]
```diff
@@ -34,103 +35,100 @@ type ActorLocation struct {
...
+			return nil, fmt.Errorf("unexpected field in AdditionalFields: %s", key)
```
Comment (gmlewis) [inline github/orgs_audit_log.go:127]:
```suggestion
			return nil, fmt.Errorf("unexpected field in AdditionalFields: %v", key)
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @WillAbides !
Just a few nits, please, then we will be ready for a second LGTM+Approval from any other contributor to this repo before merging.

**Review by gmlewis — APPROVED** [review]
Thank you, @WillAbides !
LGTM.

Awaiting second LGTM+Approval from any other contributor to this repo before merging.

**Review by valbeat — APPROVED** [review]
LGTM!

**Comment by gmlewis** [comment]
Thank you @valbeat !
Merging.

**Comment by gmlewis** [comment]
Sorry, I'm going to have to work on this one because of all the other recent merges. Hopefully tomorrow I can work on it.

**Comment by gmlewis** [comment]
Note that the new `Referrer` field added in #3032 has been moved in this PR to the `AdditionalFields` map.

**Pull request merged**